lib/tasks/salt.rake

```diff
@@ -5,6 +5,19 @@ namespace :salt do
       puts "Please provide WORKER_ID environment variable that identifies this worker"
       Process.exit 1
     end
+    if ENV["SALT_MINION_RECONCILER"] == "true"
+      # rubocop:disable Lint/HandleExceptions
+      Thread.new do
+        loop do
+          begin
+            Minion.update_grains
```

Comment:
so if this is called in a thread, and there is a possibility of multiple threads running, we might have to consider locking via mutex, as within `Minion.update_grains` we're changing the state of a shared resource. along with that we would need to have some deadlock protection as well, sth like:

```ruby
m = Mutex.new
c = ConditionVariable.new
Thread.new do
  loop do
    m.synchronize {
      begin
        c.wait(m)
        Minion.update_grains
        c.signal
      rescue StandardError
      end
      sleep 60
    }
  end
end
```

Reply:
This case is a little different IMHO because the only producer is salt, that's the only truth (it's not like two producers can be providing different inputs and we'll override the other). Moreover, since grains tend to stabilize once they change, I wouldn't see this as an issue. Also, because of the envvar the idea is to only run a reconciler on one worker (in any case, we have only one worker). For me, it's not worth the complexity of handling a condition variable. More importantly, a condition variable wouldn't help because it's local and won't protect you from other processes (if you run N workers, they are N processes, each one with the minion reconciler if you chose so), we'd need a named mutex, and as I said, I don't consider this to be an issue given how the input is produced.